fgpyo/io/io.py

```diff
@@ -0,0 +1,178 @@
+Module for reading and writing files.
+
+The functions in this module make is easy to:
+* check if a file exists and is writeable
+* check if a file and its parent directories exist and are writeable
+* check if a file exists and is readable
+* check if a path exists and is a directory
+* open an appropriate reader or writer based on the file extension
+* writitng items to a file
+* reading lines from a file
+~~~~~~~~
+Examples:
+.. code-block:: python
+   >>> import fgpyo.io.io as IO
+   >>> from pathlib import Path
+   Assert that paths exist and are readable
+   >>> path_flath: Path = Path("example.txt")
+   >>> path_compressed: Path = Path("example.txt.gz")
+   >>> IO.paths_are_readable(path_flat)
+   AssertionError: Cannot read non-existent path: example.txt
+   >>> IO.paths_are_readable(compressed_file)
+   AssertionError: Cannot read non-existent path: example.txt.gz
+   Write to and read from path
+   >>> write_lines(path = path_flat, lines_to_write=["flat file", 10])
+   >>> write_lines(path = path_compressed, lines_to_write=["gzip file", 10])
+   Read lines from paths into a list of strings
+   >>> read_lines(path = path_flat)
+   ['flat file', '10']
+   >>> read_lines(path = path_compressed)
+   ['gzip file', '10']
+"""
+
+import gzip
+import io
+import os
+from pathlib import Path
+from typing import IO
+from typing import Any
+from typing import Iterable
+from typing import Set
+from typing import TextIO
+from typing import Union
+from typing import cast
+
+COMPRESSED_FILE_EXTENSIONS: Set[str] = {".gz", ".bgz"}
+
+
+def paths_are_readable(path: Path) -> None:
+    """Checks that file exists and returns True, else raises FileNotFoundError
+
+    Args:
+        path: a Path to be investigated
+
+    Example:
+    _file_exists(path = Path("some_file.csv"))
+    """
+    assert path.exists(), f"Cannot read non-existent path: {path}"
+    assert not path.is_dir(), f"Cannot read path becasue it is a directory: {path}"
+    assert os.access(path, os.R_OK), f"Path exists but is not readable: {path}"
+
+
+def directory_exists(path: Path) -> None:
+    """Asserts one or more Paths exist and are directories
+
+    Args:
+        paths: list of one or more Paths to be investigated
+
+    Example:
+    _directory_exists(path = Path("/example/directory/"))
+    """
+    assert path.exists(), f"Path does not exist: {path}"
+    assert path.is_dir(), f"Path exists but is not a directory: {path}"
+
+
+def paths_are_writeable(path: Path, parent_must_exist: bool = True) -> None:
+    """Checks that path is writable and returns True, else raises an Exception
+    Args:
+         path: Path to be investigated
+    Example:
+    _writeable(path = Path("example.txt"))
+    """
+    assert path.exists(), f"Cannot write file because path is non-existent: {path}"
+    assert path.is_file(), f"Cannot write file because path is a directory: {path}"
+    assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+
+    if parent_must_exist:
```

Comment:
To avoid long indent blocks, consider returning early, avoiding the indents after:

```python
if not parent_must_exist:
    return None
```